### PR TITLE
Fix problems with block_area

### LINKED ISSUE #1324: Reimplement block_area with LocationIndexTree instead of graph traversal
There are several limitations with the current approach and so we should use a better node search

Discussed partially in https://github.com/graphhopper/graphhopper/pull/1306#discussion_r177076643

Failing examples likely caused by this bug: https://discuss.graphhopper.com/t/4747 and https://discuss.graphhopper.com/t/3415 and https://discuss.graphhopper.com/t/3680

Also what should happen if the area is outside of the graph - should we silently ignore this, because the area does just not contain any graph or is this a "bad request"?

### LINKED ISSUE #1905: Put back block_area for isochrone & spt
This feature got lost when the block area handling was moved out of createWeighting.

### PR DESCRIPTION
**Pull request by karussell**
This PR fixes #1324.

There were a couple of bugs with the block_area feature that are fixed within this PR:

 * virtual edges are handled now, see #1127
 * bread first search is replaced with locationIndex lookup, which avoids several problems documented in #1324
 * edgeId cache is properly used (return early!), which should lead to a speed up
 * multiple block areas are properly used
 * fixes #1905

### CODE REVIEW

File: core/src/main/java/com/graphhopper/GraphHopper.java [hunk 1]
```diff
@@ -1073,6 +1073,9 @@ else if (ALT_ROUTE.equalsIgnoreCase(algoStr))
                 return Collections.emptyList();
 
             QueryGraph queryGraph = QueryGraph.lookup(graph, qResults);
+            if (weighting instanceof BlockAreaWeighting)
+                ((BlockAreaWeighting) weighting).setQueryGraph(queryGraph);
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/GraphHopper.java]: Here we could think about a more generic solution as this is also required for the CustomWeighting in #1841.

**Comment by easbar** [comment]
Nice :) Maybe while you are at it: #1905

**Comment by karussell** [comment]
@easbar have done this now. Do you think it makes sense to repeat it or should I move it back into the createWeighting method?

**Comment by easbar** [comment]
> Do you think it makes sense to repeat it or should I move it back into the createWeighting method?

I don't have a very strong opinion here, maybe leave it as is for now?

**Pull request merged**